docs/GFXReconstruct Architectural Principles.md

```diff
@@ -0,0 +1,219 @@
+* Benefits:  
+  * It is possible to analyze an application’s API calls using the content of the capture file without the application.   The user doesn’t have to determine what calls the capture layer modified or added.  
+* Exceptions:  
+  * Because of the nature of trimming, Graphics API calls issued before the beginning of the trim range are likely to be rearranged and many omitted.  The individual calls to create objects and change object state should remain as close to the original commands issued by the application as possible.  
+
+## PRESERVE 
+
+**The graphics API calls made by the application during capture should be passed to the implementation (or downchain when layering) with as little modification as is possible.**
+
+Capture should avoid changing the content of the application’s graphics commands.  Some exceptions are necessary in order to save information required for replay or modify data so that replay is possible.
+
+* Implications:  
+  * It *may* be acceptable for the user to *request* an annotated or altered capture.  (e.g. A user may wish to save the GPU buffers used by drawing commands at capture time.  However, the preferred technique for altering the application’s graphics command stream and saving additional information for a specialization of GFXR is to make an additional upchain or downchain layer separate from GFXR. For APIs without a formal third-party layering procedure (e.g. DirectX 12), we may implement a plugin architecture for pre- or post- function and method call alteration of commands.)  
+  * Combined with the above principle “A capture file should contain as much as possible of all the graphics API calls made by the application”, a corollary is that API calls modified in capture should be stored in their original form in the capture file; a MetaData Command may encode additional information to signal to the replay application to also modify the calls on replay.  
+* Benefits:  
+  * GFXR limits the differences between running the application with and without the capture layer.  GFXR increases confidence that replay of the capture file will closely match the application’s execution without the capture layer.  
+* Exceptions:  
+  * Capture enables the Vulkan device features bufferDeviceAddressCaptureReplay, rayTracingPipelineShaderGroupHandleCaptureReplay, and accelerationStructureCaptureReplay during capture of Vulkan applications (and in replay) to allow replay on the same device and driver of ray-tracing workloads.  
```

Comment:
Perhaps add an explicit exception here of things added because of explicit communication/agreement between the app and the capture layer? The capture layer can offer additional functionality where the use of it assumes that the application (to some extent) is aware that it is debugged or captured. This includes things like debug annotations, explicit host updates, explicit frame boundaries, etc..